### ISSUE TITLE
<link rel="prefetch"> downloads aren't persisted as expected — use XHR to save

### ISSUE THREAD

**Issue by KyleAMathews** [gatsbyjs/gatsby#12800]
In Gatsby v2 we shifted to using `<link rel="prefetch">` to prefetch resources as it allows the browsers to be smarter about sharing network bandwidth between competing downloads as it can deprioritize our prefetching when there's higher priority API calls / image requests / etc.

One of our assumptions when we made this design decision was that resources we prefetch would well, stay prefetched. So we could prefetch a resource at the start of a session and if it was only used at the end of the session, the browser would keep it around.

I'd asked @addyosmani to review my [image prefetching RFC](https://github.com/gatsbyjs/rfcs/pull/35) and he mentioned prefetching can be problematic as browsers actually only keep prefetched resources around for ~5 minutes.

Which means that if a user stays on a page for 5+ minutes before clicking onto another page, the browser will need to download the next pages resources again doubling the downloaded bytes and causing extra latency when changing pages.

So we want to keep using `<link rel="prefetch">` as it's the friendliest prefetch but we need to persist the prefetched resources.

One way we could do this is to use the onload event to trigger an immediate xhr which would then make the browser persist the resource e.g. `<link rel="prefetch" as="image" href="http://example.com/image.png" onload="fetch('http://example.com/image.png')">`

The image will be already be in the memory cache so the browser will just copy it into the more permanent cache (don't remember off top of my head the names of the different browser caches).

**Comment by KyleAMathews** [comment]
A code sample:

```html
<html>
  <head>
    <link
      rel="prefetch"
      as="image"
      href="https://images.pexels.com/photos/248797/pexels-photo-248797.jpeg"
      onload="fetch('https://images.pexels.com/photos/248797/pexels-photo-248797.jpeg')"
    />
  </head>
  <body>
    <h2>Just testing a trick</h2>
  </body>
</html>
```

**Comment by addyosmani** [comment]
So the logic here is..

- Prefetch URL with a low priority 
- When the resource is prefetched, force a fetch() which consumes the resource and persists it to the browser cache based on cache-control rules 
- Everything prefetched is now persisted under the assumption it will be needed at some point 

If we can do further analysis on how well this holds up, it might work. cc @yoavweiss as a sanity check.

**Comment by yoavweiss** [comment]
A few questions:
* Are the resources in question cacheable? 
   - If so, they will (or at least should) stay in cache beyond the 5 minutes window. That [window](https://cs.chromium.org/chromium/src/net/http/http_cache_transaction.cc?type=cs&q=LOAD_PREFETCH&sq=package:chromium&g=0&l=2761) makes sure that non-cacheable (but storable) resources in cache can escape validation for the first use during that time. After that time, or in followup uses, they follow the regular revalidation rules, which means that if they are cacheable and don't require validation, they should not trigger requests even if the user lingers on the page beyond the 5 minutes.
   - Have you seen instances where these resources are being refetched if the user lingers on the prefetching page?
* What is your use-case for prefetching images?
   - We are currently [in the process](https://github.com/whatwg/html/pull/4115) of removing prefetch support for subresources, as it introduces a lot of complexity when interacting with Service Workers, as well as introduces privacy issues when considering double-keyed caches.
   - If you have strong use-cases that will justify maintaining those use-cases in some way (e.g. only for same origin, by providing markup indicating the destination SW scope, etc), it would be great to hear it.
* I'd caution against heavily relying on prefetch's prioritization downgrade too much, as it [doesn't always prevent contention](https://bugs.chromium.org/p/chromium/issues/detail?id=651160) with resources for the current page, especially if their discovery is delayed. It might be better to inject those prefetches only after you know that the current page is fully rendered and interactive.

**Comment by wardpeet** [comment]
Thanks for all the information! I agree with @yoavweiss on `What is your use-case for prefetching images? unsure if it's a big deal to show op placeholder images on navigation? Also won't service worker be able to save it to cache?

**Comment by KyleAMathews** [comment]
We want to prefetch images so they show up sooner :-)

When you click between pages on a Gatsby site, most of the page renders immediately but images don't as they're not discovered until the page is rendered. We want to add the ability to Gatsby to mark images as "critical" so they we can prefetch them. A good example of images loading late is this recently launched Gatsby site https://www.prima.co — try clicking between pages and you get a blinking effect when images load instead of a nice smooth single render.

This RFC I posted is related: https://github.com/gatsbyjs/rfcs/pull/35

>Have you seen instances where these resources are being refetched if the user lingers on the prefetching page?

I haven't seen instances of this (though I also haven't been looking) but if prefetching does follow cache-control headers then we're actually probably fine.

>I'd caution against heavily relying on prefetch's prioritization downgrade too much, as it doesn't always prevent contention with resources for the current page

Hmmm that's disappointing...

>It might be better to inject those prefetches only after you know that the current page is fully rendered and interactive.

Yeah we can do this — we do add some prefetches to the `<head>` of our rendered HTML so those would start loading immediately but we can defer those instead to start after the page is interactive.

**Comment by yoavweiss** [comment]
> We want to prefetch images so they show up sooner :-)
> 
> When you click between pages on a Gatsby site, most of the page renders immediately but images don't as they're not discovered until the page is rendered. We want to add the ability to Gatsby to mark images as "critical" so they we can prefetch them. A good example of images loading late is this recently launched Gatsby site https://www.prima.co — try clicking between pages and you get a blinking effect when images load instead of a nice smooth single render.

So the problems with subresource prefetching are:
* We're not sure which Service Worker scope will handle those subresources.
* In a double-keyed cache model, those prefetches can leak resources between different origins.

Are you actually navigating between those pages? Or are those "soft navigations"?
If it's the latter, you can (dynamically) preload those images, as they are destined for the current navigation.

Otherwise, I guess we can make subresource prefetches that are destined for same origin with same SW scope work. Would require some opt-in though, so I want to make sure the use-case is there.
